src/op_analytics/cli/subcommands/pulls/agora/delegate_events.py

```diff
@@ -0,0 +1,161 @@
+from dataclasses import dataclass
+from op_analytics.cli.subcommands.pulls.agora.data_access import (
+    Agora,
+    write,
+    _camelcase_to_snakecase,
+    parse_isoformat_with_z,
+)
+from op_analytics.coreutils.logger import structlog
+from op_analytics.coreutils.request import get_data
+from op_analytics.coreutils.threads import run_concurrently
+
+from typing import Any
+import requests
+import requests.exceptions
+import pandas as pd
+import os
+
+log = structlog.get_logger()
+
+BASE_URL = "https://vote.optimism.io/api/v1"
+DELEGATES_ENDPOINT = f"{BASE_URL}/delegates"
+API_KEY = os.environ["AGORA_API_KEY"]
+session = requests.Session()  # Create a session object
+
+
+@dataclass
+class AgoraDelegateVotes:
+    votes_df: pd.DataFrame
+
+
+@dataclass
+class PaginatedResponse:
+    has_next: bool
+    next_offset: int
+    data: Any
+
+
+@dataclass
+class SimplePaginator:
+    url: str
+    limit: int = 50
+
+    def request(self, offset):
+        result = get_data(
+            session,
+            url=self.url,
+            headers={"Authorization": f"Bearer {API_KEY}"},
+            params={"offset": offset, "limit": self.limit},
+        )
+
+        assert "meta" in result
+        assert "data" in result
+
+        return PaginatedResponse(
+            has_next=result["meta"]["has_next"],
+            next_offset=result["meta"]["next_offset"],
+            data=result["data"],
+        )
+
+    def fetch_all(self):
+        offset = 0
+        all_data = []
+
+        while True:
+            response = self.request(offset)
+            all_data.extend(response.data)
+
+            if not response.has_next:
+                break
+
+            offset = response.next_offset
+
+        return all_data
+
+
+def fetch_event_data(delegates: list, batch_size: int, endpoint: str, workers: int) -> pd.DataFrame:
+    failed_addresses = []
+
+    def _fetch_with_fallback(address: str, endpoint: str) -> list:
+        paginator = SimplePaginator(url=f"{BASE_URL}/delegates/{address}/{endpoint}", limit=50)
+        try:
+            return paginator.fetch_all()
+        except requests.exceptions.HTTPError as e:
+            if e.response.status_code == 500:
+                log.error(f"Internal Server Error for address: {address}, URL: {e.response.url}")
+            else:
+                log.error(
+                    f"HTTP Error for address: {address}, URL: {e.response.url}, Status Code: {e.response.status_code}"
+                )
+            failed_addresses.append(address)
+            return []
+
+    all_delegate_data = []
+
+    for i in range(0, len(delegates), batch_size):
```

Comment:
Using batches when we already parallelized with threads seems unnecessary.  

Also, instead of a second pass of retrying, my recommendation would be to decorate `_fetch_with_fallback` with `@stamina.retry` and that way you can have the threads themselves take care of retries. 

If you want to tolerate some failures, because retries are not enough, take a look at `run_concurrently_store_failures` that seems better than catching exceptions like we are doing in `_fetch_with_fallback`.